command/agent/cache/api_proxy.go

```diff
@@ -36,7 +38,16 @@ func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse,
 		return nil, err
 	}
 	client.SetToken(req.Token)
-	client.SetHeaders(req.Request.Header)
+
+	// http.Transport will transparently request gzip and decompress the response, but only if
+	// the client doesn't manually set the header (in which case the client has to handle
+	// the decompression. If gzip has already been set, remove it to avoid triggering the manual
+	// handling requirement.
+	h := clone(req.Request.Header)
```

Comment:
Why clone?

Why not use the Header built-in functions to handle removal?

Reply:
`clone` b/c I didn't think this Send function should mutate the caller's request object. If that's not an issue, then we can get rid of it.

Is there a built-in to remove just a single value? https://golang.org/pkg/net/http/#Header.Del will delete all values.

Reply:
Unless the request object is being used for some other purpose down the wire I think it's fine to mutate it. I would think once it gets to this point it won't be further modified and we don't log request headers (or if we do it's before this, I would expect).

I would assume you'd have the same problem with `deflate` or any other encoding so I imagine you should remove them all. You could first cache them if you wanted to then use it to decide how to respond to the client, which is not a bad option, since the client is telling the cache what it will accept. But in that case you'd probably want to cache them all, not remove gzip and then cache the rest.

Reply:
`gzip` is a special case because the Go std library transparently adds it depending on headers being set. Nonetheless, I think it's probably fine to just unconditionally remove the `Accept-Encoding` header on the request object. We're not giving up gzip to Vault since the agent's Transport will add it in anyway. I think caching can be deferred until we need to actually reprocess the response (which seems unlikely, since we don't and probably shouldn't compress Vault data responses).